### PR TITLE
Add GitLab Application deployment using Crossplane Resources example

### PR DESCRIPTION
**Pull request by ichekrygin**
image

 Resolves #

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make generate`) has been run to update object specifications, if necessary.
- [ ] CRD manifests generation (`make manifests`) has been run to update CRD manifests yaml file specifications, if necessary.
- [ ] Update dependencies (`make vendor`) has been run to update `Gopkg.lock` file, if necessary
- [ ] RBAC permissions in `clusterrole.yaml` have been updated to include new types, if necessary
- [ ] All related commits have been squashed to improve readability.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

### CODE REVIEW

File: docs/gitalb/gitlab-gcp.md [hunk 1]
```diff
@@ -0,0 +1,380 @@
+# Deploying GitLab in GCP
```
Comment (negz) [inline docs/gitalb/gitlab-gcp.md]: This file is in directory `gitalb`, not `gitlab`.

File: docs/gitalb/gitlab-gcp.md [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+gitlab-lfs             Opaque                                4      7m4s
+gitlab-packages        Opaque                                4      2m28s
+gitlab-postgresql      Opaque                                3      30m
+gitlab-pseudonymizer   Opaque                                4      7m2s
+gitlab-redis           Opaque                                1      28m
+gitlab-registry        Opaque                                4      7m1s
+gitlab-uploads         Opaque                                4      7m1s
+```
+
+Note: Kubernetes cluster claim is created in "privileged" ` mode; thus the kubernetes cluster resource secret is located in `crossplane-system` namespace, however, will not need to use this secret for our GitLab demo deployment.
```
Comment (negz) [inline docs/gitalb/gitlab-gcp.md]: I think there's a rogue code quote mark (`) here.

File: docs/gitalb/gitlab-gcp.md [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+    --clusterrole cluster-admin \
+    --user [your-gcp-user-name]
+```
+
+##### Export and Convert Secrets
+GitLab requires to provide connection information in the specific format per cloud provider.
+In addition, we need to extract endpoints and additional managed resource properties and add them to helm values.
+
+We will use a convenience script for this purpose.
+Note: you output may be different
```
Comment (jbw976) [inline docs/gitalb/gitlab-gcp.md]: typo: `your output`

File: docs/gitalb/gitlab-gcp.md [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+```bash
+kubectl delete -Rf cluster/examples/gitlab/gcp/resource-classes/
+---
+resourceclass.core.crossplane.io "standard-gcp-bucket" deleted
+resourceclass.core.crossplane.io "standard-gcp-cluster" deleted
+resourceclass.core.crossplane.io "standard-gcp-postgres" deleted
+resourceclass.core.crossplane.io "standard-gcp-redis" deleted
+```
+
+Delete gcp-provider and secrets
```
Comment (jbw976) [inline docs/gitalb/gitlab-gcp.md]: should we have a step that tells them to wait for all the external resources to get cleaned up before deleting the provider objects?

File: gitlab-gcp.yaml [hunk 5]
```diff
@@ -0,0 +1,4795 @@
...
+  clusterIP: ""
+  ports:
+    - name: stats
+      port: 18080
+      targetPort: stats
+  selector:
+    app: nginx-ingress
+    component: "controller"
+    release: gitlab-demo
+  type: "Cl
```
Comment (jbw976) [inline gitlab-gcp.yaml]: what are all these blank entries? are the indicative of a problem? should they be removed?

File: cluster/examples/gitlab/gcp/resource-claims/buckets/artifacts.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+apiVersion: storage.crossplane.io/v1alpha1
+kind: Bucket
+metadata:
+  name: gitlab-artifacts
+  namespace: default
+spec:
+  classReference:
+    name: standard-gcp-bucket
+    namespace: crossplane-system
+  name: gitlab-demo-artifacts
```
Comment (negz) [inline cluster/examples/gitlab/gcp/resource-claims/buckets/artifacts.yaml]: I wonder whether we should somehow handle the fact that these names are globally unique - add a prefix via a script or something.

File: cluster/examples/gitlab/gcp/resource-classes/bucket.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# ResourceClass that defines the blueprint for how a "standard" GCP bucket
+# should be dynamically provisioned
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceClass
+metadata:
+  name: standard-gcp-bucket
+  namespace: crossplane-system
+parameters:
+  bucketPolicyOnly: "true"
+  labels: "app:gitlab-demo"
```
Comment (negz) [inline cluster/examples/gitlab/gcp/resource-classes/bucket.yaml]: we use `app:` here but `application:` for Kubernetes.

File: gitlab-gcp.yaml (hunk 5 above)
Comment (ichekrygin) [reply gitlab-gcp.yaml]: Those are generated by `helm template` and most likely due to optional or "not configured" components. This demo does not modify either the GitLab helm chart nor rendered yaml file.

File: docs/gitalb/gitlab-gcp.md (hunk 1 above)
Comment (ichekrygin) [reply docs/gitalb/gitlab-gcp.md]: :heavy_check_mark:

File: docs/gitalb/gitlab-gcp.md (hunk 2 above)
Comment (ichekrygin) [reply docs/gitalb/gitlab-gcp.md]: :heavy_check_mark:

File: cluster/examples/gitlab/gcp/resource-claims/buckets/artifacts.yaml (hunk 6 above)
Comment (ichekrygin) [reply cluster/examples/gitlab/gcp/resource-claims/buckets/artifacts.yaml]: I think that is a great idea for follow up PR.

Comment (ichekrygin) [reply cluster/examples/gitlab/gcp/resource-claims/buckets/artifacts.yaml]: https://github.com/crossplaneio/crossplane/issues/431

File: docs/gitalb/gitlab-gcp.md [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+gitlab-gke   Bound    standard-gcp-cluster   gke-af012df6-6e2a-11e9-ac37-9cb6d08bde99   71m
+```
+- Using `CLUSTER-REF` get GKECluster resource
+```bash
+kubectl get gkecluster gke-af012df6-6e2a-11e9-ac37-9cb6d08bde99 -n crossplane-system
+---
+NAME                                       STATUS   STATE     CLUSTER-NAME                               ENDPOINT          CLUSTER-CLASS          LOCATION        RECLAIM-POLICY   AGE
+gke-af012df6-6e2a-11e9-ac37-9cb6d08bde99   Bound    RUNNING   gke-af11dfb1-6e2a-11e9-ac37-9cb6d08bde99   130.211.208.249   standard-gcp-cluster   us-central1-a   Delete           72m
+```
+- Record the `CLUSTER_NAME` value
```
[the remark this answers is not in the document]
Comment (ichekrygin) [reply docs/gitalb/gitlab-gcp.md]: I think this is a great suggestion for the follow-up PR

File: docs/gitalb/gitlab-gcp.md [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+If `helm fetch` command is successful, you should see a new file created in your CWD:
+```bash
+ls -l external-dns-*
+---
+-rw-r--r-- 1 user user 8913 May  3 23:24 external-dns-1.7.5.tgz
+```
+
+- Render Helm chart into `yaml`, set values and apply to your GKE cluster
+```bash
+helm template external-dns-1.7.5.tgz --name gitlab-demo --namespace kube-system  \
```
[the remark this answers is not in the document]
Comment (ichekrygin) [reply docs/gitalb/gitlab-gcp.md]: I think this is a great suggestion for the follow-up PR

File: docs/gitalb/gitlab-gcp.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (ichekrygin) [reply docs/gitalb/gitlab-gcp.md]: https://github.com/crossplaneio/crossplane/issues/432

File: docs/gitalb/gitlab-gcp.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (ichekrygin) [reply docs/gitalb/gitlab-gcp.md]: https://github.com/crossplaneio/crossplane/issues/432

File: .gitignore [hunk 10]
```diff
@@ -6,3 +6,11 @@
 cover.out
 /vendor
 /.vendor-new
+
+# gitlab example
+cluster/examples/gitlab/gcp/values-psql.yaml
```
Comment (bassam) [inline .gitignore]: nit: creating a local `.gitignore` file in `cluster/examples/gitlab` would put these rules closer to the code using them.

File: docs/gitlab/gitlab-gcp.md [hunk 11]
```diff
@@ -0,0 +1,479 @@
...
+
+Add your user account to the cluster admin role
+```bash
+kubectl create clusterrolebinding cluster-admin-binding \
+    --clusterrole cluster-admin \
+    --user [your-gcp-user-name]
+```
+
+#### External DNS
+- Fetch [Eternal-DNS](https://github.com/helm/charts/tree/master/stable/external-dns) helm chart
```
Comment (bassam) [inline docs/gitlab/gitlab-gcp.md]: nit: typo. `External-DNS`

File: docs/gitlab/gitlab-gcp.md [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+#### Managed Resource Secrets
+Decide on the GKE cluster namespace where GitLab application artifacts will be deployed.
+
+We will use: `gitlab`, and for convenience we will [set our current context](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#setting-the-namespace-preference) to this namespace 
+```bash
+kubectl create ns gitlab
+kubectl config set-context $(kubectl config current-context) --namespace=gitlab
+```
+
+##### Export and Convert Secrets
```
Comment (bassam) [inline docs/gitlab/gitlab-gcp.md]: would be great to highlight how we solve this problem going forward, and point an issue.

**Comment by bassam** [comment]
Nice work! Does this assume that gitlab and crossplane run in the same cluster, or are we operating on a remote GKE cluster? I think both are interesting.

File: .gitignore (hunk 10 above)
Comment (ichekrygin) [reply .gitignore]: :heavy_check_mark:

File: docs/gitlab/gitlab-gcp.md (hunk 11 above)
Comment (ichekrygin) [reply docs/gitlab/gitlab-gcp.md]: :heavy_check_mark:

File: docs/gitlab/gitlab-gcp.md (hunk 12 above)
Comment (ichekrygin) [reply docs/gitlab/gitlab-gcp.md]: :heavy_check_mark:

**Comment by ichekrygin** [comment]
As written, this example will create a new GKE Kubernetes cluster for GitLab deployment.
I created https://github.com/crossplaneio/crossplane/issues/436 to track GKE in-cluster example.

**Pull request merged**